### PR TITLE
Grafter

### PR DESCRIPTION
**Pull request by Bam4d**
Single agent implementation of training grafter agents.

they appear to do very well: 
image

### CODE REVIEW

**Comment by vwxyzjn** [comment]
What’s a grafter agent?

**Comment by Bam4d** [comment]
> What’s a grafter agent?

Griddly version of crafter == grafter

**Comment by Bam4d** [comment]
ezgif-3-bcaddb8a9c

File: enn_zoo/enn_zoo/griddly_env/level_generators/crafter_generator.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        self._num_players = num_players
+        self._walkable = {"G", "P", "S"}
+        self._random = np.random.RandomState(seed)
+
+    def _get_level_string(self, world: np.ndarray) -> str:
+        level_string = ""
+        for h in range(0, self._height):
+            for w in range(0, self._width):
+                level_string += world[w, h].ljust(4)
+            level_string += "\n"
```
Comment (cswinter) [inline enn_zoo/enn_zoo/griddly_env/level_generators/crafter_generator.py:27]: this is O(n^2), faster way is to create a list with all the string firsts and then use `"".join`

**Review by cswinter — APPROVED** [review]
LGTM!
Only think I found is an O(n^2) in environment creation, fixing that could make reset faster.

**Comment by cswinter** [comment]
@Bam4d I'm planning to land some final refactors and split out entity-gym soon, I recommend merging this PR and then adding any remaining changes later if necessary to simplify things.

**Pull request merged**